### DISCUSSION TITLE
I figured out how to get Instance Segmenation color of vehicles

### DISCUSSION THREAD

**Discussion by Emcyz** [carla-simulator/carla#5047] [Show and tell]
in carla/Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Game/Tagger.cpp
there is newly added code at line 70
![image](https://user-images.githubusercontent.com/11612272/148172616-9badeb3b-2500-4667-a78e-ab1972a2f634.png)

I thought this code determines color of instance segmantaion
but Actor.GetUniqueID() is not Actor Id in carla simulator
this function is in Unreal Engine which returns Unique ID in Unreal Engine(I guess), not Id in carla
but I could't replace this code to use carla Actor Id
Instead, I changed carla Actor Id to Unique ID

in carla/Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Actor/ActorRegistry.cpp at line 80
![asdf](https://user-images.githubusercontent.com/11612272/148173827-64d0c24e-af70-4054-b320-048a5df78006.png)
there is code that determines Id of actor.
IdType Id = ++FActorRegistry::ID_COUNTER;

I removed it and replace with
IdType Id = Actor.GetUniqueID();
then make PythonAPI and I test it to see whether it really works

I run manual_control.py with
print(world.player.id)
print((world.player.id & 0x00ff) >> 0)
print((world.player.id & 0xff00) >> 8)

![InkedInkedqwer_LI](https://user-images.githubusercontent.com/11612272/148175509-627851fb-a593-4447-b155-1f4e32b77810.jpg)
then I captured screen, checked the RGB value with OpenCV and it was correct!!
ID is 116793, G is 57, B is 200

I hope this can help other guys  :)

**Comment by DennisKlpng** [comment]
Great find!


We don't know yet if there is any way to get the Unreal-Engine ID for a Carla Actor with a specific Carla ID or is there?

The latter would be especially useful for anyone who can't / doesn't want to build Carla from source

**Comment by eskildbr** [comment]
Hi, a quick question. Did you reverse-engineer or something like that to find the ID when using OpenCV to get the RGB values?

**Reply by Emcyz** [reply]
I used cv2.imshow(windowname, image)
If image has three channel(RGB, not RGBA), you can zoom that image and see RGB value like the picture I posted above

**Reply by eskildbr** [reply]
The RGB values show (10, 57, 200) by using cv2 right? but the ID in Carla is 116793. How did you find out that the RGB values 10, 57, 200 corresponds to the ID? Or did cv2 give the RGB value (116793, 57, 200)?

Basically, how did you match the IDs from Carla and the RGB value?

**Reply by eskildbr** [reply]
Figured it out :D I took the green value << 8, plus the blue value and found the ID. 

(green value << 8) + blue value = ID

**Reply by gogokotsev00** [reply]
Actually, the correct equation is (blue value << 8) + green value = ID

**Reply by JosenJiang** [reply]
> Actually, the correct equation is (blue value << 8) + green value = ID

Sorry, I am still a little confused. According to the equation you gave, the result is not equal to ID. Is there something wrong?
![image](https://user-images.githubusercontent.com/109076997/226280187-6c862acb-3411-458e-9d3b-cdcc67b78bce.png)

**Reply by gogokotsev00** [reply]
> > Actually, the correct equation is (blue value << 8) + green value = ID
> 
> Sorry, I am still a little confused. According to the equation you gave, the result is not equal to ID. Is there something wrong? ![image](https://user-images.githubusercontent.com/109076997/226280187-6c862acb-3411-458e-9d3b-cdcc67b78bce.png)

Yes, an important discovery that I made during my adventure with CARLA (and forgot to mention here :D) was that there is a `boundary value = 65536` at which an overload appears. This happens due to the fact that we have 8 bits for each of the R, G and B channels, where the max. value of these 8 bits is 1111 1111<sub>2</sub> = 255<sub>10</sub> and for 256<sub>10</sub>=1 0000 0000<sub>2</sub> we have the overload. If we use the SHIFT LEFT operator, on 256<sub>10</sub> we'll then get exactly 65536<sub>10</sub>. The `(B<<8) + G` value can be max. `65535 = 255<<8 + 255`, which is not the case for the actor's ID. Therefore, in order to know if the calculated result actually does match the actor's ID, I used the following code: 
```
    boundary_value = 65536
    if actor_id > boundary_value:
        actor_id -= math.floor(actorr_id / boundary_value) * boundary_value
   if (B<<8) + G == actor_id:
        return true
```

At the end of this code snippet, you'll receive the modified actor's ID that should match the  `(B<<8) + G` result, otherwise the pixel does not belong to this specific actor.

**Reply by JosenJiang** [reply]
Thanks a lot! But according to your method, there may be different actor_ids for the same (B, G) value. 
For example:
```
actor_id_1 = x (x < 65535)
actor_id_2 = x + 65535
```
I don't know how the UE4 registers an ID for an actor. I am confused whether it will happen.
We can just decode the actor_id to get instance segmentation color. But we can not decode instance segmentation color to get the actor_id.

I try to use A channel like this:
![image](https://user-images.githubusercontent.com/109076997/226521574-00cfd58e-5954-4a27-a682-5012d40395c7.png)
But all values in A channel I get is 255. I have not found solutions.

**Reply by gogokotsev00** [reply]
> Thanks a lot! But according to your method, there may be different actor_ids for the same (B, G) value. For example:
> 
> ```
> actor_id_1 = x (x < 65535)
> actor_id_2 = x + 65535
> ```
> 
> I don't know how the UE4 registers an ID for an actor. I am confused whether it will happen. We can just decode the actor_id to get instance segmentation color. But we can not decode instance segmentation color to get the actor_id.
> 
> I try to use A channel like this: ![image](https://user-images.githubusercontent.com/109076997/226521574-00cfd58e-5954-4a27-a682-5012d40395c7.png) But all values in A channel I get is 255. I have not found solutions.

Yes, when there are identical actor IDs you have to see the red channel and decide whether this is the type of actor you need. Channel A can be omitted for this task, it does not play a role in it. 

I couldn't understand what you mean by decoding the instance segmentation colour, probably that we cannot get the correct actor ID from the colours, well that's correct, I also don't know if there is a way.

**Reply by JosenJiang** [reply]
Yes, that's what I mean. In order to get the correct actor ID from the colors, I try to use G, B and A channels to encode an ID greater than 65535. But I always get 255 from A channel.
Thanks any way.

**Comment by Mofeed-Chaar** [comment]
many thanks for your information.
my question is, how could I get a Unique id using Python code?????

**Reply by Gowtham-Raj-M** [reply]
Hi @Mofeed-Chaar, Did you find solution to get a Unique id using Python code?

**Reply by Mofeed-Chaar** [reply]
not yet
i will try to find a way to run a code of c++ inside a python code

**Reply by Gowtham-Raj-M** [reply]
Hi @Mofeed-Chaar , Actually i tried different approaches to get color code for objects in Carla, but not able to get the value. Then i tried with above mentioned approach with carla built from source , now i am getting color code for each object.

**Reply by DennisKlpng** [reply]
Hi, I didn't find any solution to get it directly, what you can do if you can't build from source and if you can do some sort of post-processing (so, depends on your use case):
* move all actors away, afterwards move each actor (by Carla ID) separately directly in front of one instance segementation camera (setting velocity to 0, deactivating physics is beneficial)
* take one frame, save the image + the actor of the player you moved, move the actor away again
* Repeat until you have one instance segmented image for each actor
When counting the number of pixels with the correct actor class, you should have an association between color and Carla ID for this simulation run

Doesn't work for every use case, but when you just record data and can do post-processing it was reliable enough for me

**Comment by venitws2** [comment]
> in carla/Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Game/Tagger.cpp there is newly added code at line 70 ![image](https://user-images.githubusercontent.com/11612272/148172616-9badeb3b-2500-4667-a78e-ab1972a2f634.png)
> 
> I thought this code determines color of instance segmantaion but Actor.GetUniqueID() is not Actor Id in carla simulator this function is in Unreal Engine which returns Unique ID in Unreal Engine(I guess), not Id in carla but I could't replace this code to use carla Actor Id Instead, I changed carla Actor Id to Unique ID
> 
> in carla/Unreal/CarlaUE4/Plugins/Carla/Source/Carla/Actor/ActorRegistry.cpp at line 80 ![asdf](https://user-images.githubusercontent.com/11612272/148173827-64d0c24e-af70-4054-b320-048a5df78006.png) there is code that determines Id of actor. IdType Id = ++FActorRegistry::ID_COUNTER;
> 
> I removed it and replace with IdType Id = Actor.GetUniqueID(); then make PythonAPI and I test it to see whether it really works
> 
> I run manual_control.py with print(world.player.id) print((world.player.id & 0x00ff) >> 0) print((world.player.id & 0xff00) >> 8)
> 
> ![InkedInkedqwer_LI](https://user-images.githubusercontent.com/11612272/148175509-627851fb-a593-4447-b155-1f4e32b77810.jpg) then I captured screen, checked the RGB value with OpenCV and it was correct!! ID is 116793, G is 57, B is 200
> 
> I hope this can help other guys :)

I think this is a very easy to implement solution for the people who are currently working on the CARLA simulator and using the instance segmentation camera because it just replaces a single line of code. However, the reason why the CARLA team still did not solve this issue by using this method might be due to changing the ID assignment mechanism completely. Anyway, I used this method and it solved my problem. Keep in mind that you need to rebuild both the server and client sides from the source after doing this.
